### PR TITLE
feat: add an example of using Booleans for eager/lazy lists

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by mlutze** [comment]
Why `HList`?

**Comment by magnus-madsen** [comment]
> Why `HList`?

Temporary name. Suggestions most welcome :)

File: examples/type-level-programming-eager-lazy-list.flix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+/// A hybrid zip function which is eager if both arguments are eager.
+/// If either argument (or both) is lazy then the list is lazy.
+def zip(l1: LEList[a, e1], l2: LEList[b, e2]): LEList[(a, b), e1 and e2] =
+    typematch (BoolProxy.BoolProxy : BoolProxy[e1 and e2]) {
+        case _: (BoolProxy[true]) =>
+            unchecked_cast(zipE(l1, l2) as LEList[(a, b), e1 and e2])
+        case _: _                 =>
+            unchecked_cast(zipL(l1, l2) as LEList[(a, b), e1 and e2])
```
Comment (mlutze) [inline examples/type-level-programming-eager-lazy-list.flix:42]: Really unfortunate that we have these casts :(

**Review by mlutze — APPROVED** [review]

File: examples/type-level-programming-eager-lazy-list.flix (hunk 1 above)
Comment (magnus-madsen) [reply examples/type-level-programming-eager-lazy-list.flix:42]: Just add GADTs? :P

Comment (magnus-madsen) [reply examples/type-level-programming-eager-lazy-list.flix:42]: Actually that may not be enough. The casts are part of the specification no?

Comment (mlutze) [reply examples/type-level-programming-eager-lazy-list.flix:42]: If we get a `return SomeType` clause like in Coq GADTs then it should be enough.

**Pull request merged**